### PR TITLE
Add chat model sample for gemma models

### PR DESCRIPTION
**Pull request by olpipi**
CVS-133845

### CODE REVIEW

File: text_generation/causal_lm/cpp/CMakeLists.txt [hunk 1]
```diff
@@ -28,3 +28,11 @@ find_package(OpenVINO REQUIRED COMPONENTS Runtime)
...
 set_target_properties(speculative_decoding_lm PROPERTIES CXX_STANDARD 17)
 set_target_properties(speculative_decoding_lm PROPERTIES CXX_STANDARD_REQUIRED ON)
+
+add_executable(chat_model_lm chat_model_lm.cpp)
+target_compile_definitions(chat_model_lm PRIVATE OPENVINO_TOKENIZERS_PATH=\"$<TARGET_FILE:openvino_tokenizers>\")
+target_include_directories(chat_model_lm PRIVATE ./)
+find_package(OpenVINO REQUIRED COMPONENTS Runtime)
+target_link_libraries(chat_model_lm PRIVATE openvino::runtime)
+set_target_properties(chat_model_lm PROPERTIES CXX_STANDARD 17)
+set_target_properties(chat_model_lm PROPERTIES CXX_STANDARD_REQUIRED ON)
```
Comment (ilya-lavrenov) [inline text_generation/causal_lm/cpp/CMakeLists.txt:38]: could you please also add GHA CI pipeline for this sample?

File: text_generation/causal_lm/cpp/chat_model_lm.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    inp.set_shape({BATCH_SIZE, tokens.size()});
+    for (size_t idx = 0; idx < tokens.size(); ++idx) {
+        inp.data<int64_t>()[idx] = tokens.at(idx);
+    }
+    detokenizer.infer();
+    return detokenizer.get_output_tensor().data<std::string>()[0];
+}
+}  // namespace
+
+std::string generate_chat_promt_gemma(const std::string& input) {
```
Comment (ilya-lavrenov) [inline text_generation/causal_lm/cpp/chat_model_lm.cpp:29]:
```suggestion
std::string generate_chat_prompt_gemma(const std::string& input) {
```

Comment (ilya-lavrenov) [inline text_generation/causal_lm/cpp/chat_model_lm.cpp:29]: it should be in unnamed namespace

File: text_generation/causal_lm/cpp/chat_model_lm.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    // The model can be compiled for GPU as well
+    ov::InferRequest lm =
+        core.compile_model(std::string{argv[1]} + "/openvino_model.xml", "CPU").create_infer_request();
+
+    int64_t total_positions = 0;
+    int32_t global_beam_idx = 0;
+
+    std::string promt;
+    for (;;) {
+        std::cout << "User promt:\n";
```
Comment (ilya-lavrenov) [inline text_generation/causal_lm/cpp/chat_model_lm.cpp:56]:
```suggestion
        std::cout << "User prompt:\n";
```

File: text_generation/causal_lm/cpp/chat_model_lm.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        std::getline(std::cin, promt);
+        std::cout << "\n";
+
+        if (!promt.compare("Stop!"))
```
Comment (ilya-lavrenov) [inline text_generation/causal_lm/cpp/chat_model_lm.cpp:60]: Looks like we need to print somewhere at the beginning, that `Stop!` is a keyword.

File: text_generation/causal_lm/cpp/chat_model_lm.cpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        auto [input_ids, new_attention_mask] = tokenize(tokenizer, std::move(promt));
+
+        // Initialize inputs
+        lm.set_tensor("input_ids", input_ids);
+
+        auto attention_mask = lm.get_tensor("attention_mask");
+        // attention to all previous tokens + new tokens
+        ov::Shape mask_shape{1, attention_mask.get_shape().at(1) + new_attention_mask.get_shape().at(1)};
+        attention_mask.set_shape(mask_shape);
+        std::fill_n(attention_mask.data<int64_t>(), ov::shape_size(mask_shape), 1);
```
Comment (ilya-lavrenov) [inline text_generation/causal_lm/cpp/chat_model_lm.cpp:74]:
```suggestion
        std::fill_n(attention_mask.data<int64_t>(), attention_mask.get_size(), 1);
```

File: text_generation/causal_lm/cpp/chat_model_lm.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+        auto position_ids = lm.get_tensor("position_ids");
+        position_ids.set_shape(input_ids.get_shape());
+        auto check = attention_mask.get_shape().at(1);
```
Comment (ilya-lavrenov) [inline text_generation/causal_lm/cpp/chat_model_lm.cpp:78]: it seems unused.

File: text_generation/causal_lm/cpp/chat_model_lm.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            if (next_tokens.empty()) {
+                break;
+            }
+            size_t batch_size = next_tokens.size();
+            // Set pointers
+            lm.set_tensor("input_ids", ov::Tensor{ov::element::i64, {batch_size, 1}, next_tokens.data()});
+            lm.set_tensor("beam_idx", ov::Tensor{ov::element::i32, {batch_size}, next_beams.data()});
+            // Set auxiliary inputs
+            ov::Tensor attention_mask = lm.get_tensor("attention_mask");
+            ov::Shape mask_shape{batch_size, attention_mask.get_shape().at(1) + 1};
```
Comment (ilya-lavrenov) [inline text_generation/causal_lm/cpp/chat_model_lm.cpp:102]: maybe we can define `seq_len_dim_idx = 1` and use:
```suggestion
            ov::Shape mask_shape{batch_size, attention_mask.get_shape().at(seq_len_dim_idx) + 1};
```

File: text_generation/causal_lm/cpp/chat_model_lm.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    // Compile models
+    ov::Core core;
+    core.add_extension(OPENVINO_TOKENIZERS_PATH);  // OPENVINO_TOKENIZERS_PATH is defined in CMakeLists.txt
+    // tokenizer and detokenizer work on CPU only
+    ov::InferRequest tokenizer =
+        core.compile_model(std::string{argv[1]} + "/openvino_tokenizer.xml", "CPU").create_infer_request();
+    ov::InferRequest detokenizer =
+        core.compile_model(std::string{argv[1]} + "/openvino_detokenizer.xml", "CPU").create_infer_request();
+    // The model can be compiled for GPU as well
```
Comment (ilya-lavrenov) [inline text_generation/causal_lm/cpp/chat_model_lm.cpp:47]: have you tried it on GPU?

File: text_generation/causal_lm/cpp/CMakeLists.txt (hunk 1 above)
Comment (ilya-lavrenov) [reply text_generation/causal_lm/cpp/CMakeLists.txt:38]: maybe we can also move this sample to a dedicated folder `llm-chatbot` with its own README.md?

File: text_generation/causal_lm/cpp/chat_model_lm.cpp (hunk 2 above)
Comment (olpipi) [reply text_generation/causal_lm/cpp/chat_model_lm.cpp]: done

Comment (olpipi) [reply text_generation/causal_lm/cpp/chat_model_lm.cpp]: done

File: text_generation/causal_lm/cpp/chat_model_lm.cpp (hunk 3 above)
Comment (olpipi) [reply text_generation/causal_lm/cpp/chat_model_lm.cpp]: done

File: text_generation/causal_lm/cpp/chat_model_lm.cpp (hunk 4 above)
Comment (olpipi) [reply text_generation/causal_lm/cpp/chat_model_lm.cpp]: done

File: text_generation/causal_lm/cpp/chat_model_lm.cpp (hunk 5 above)
Comment (olpipi) [reply text_generation/causal_lm/cpp/chat_model_lm.cpp]: done

File: text_generation/causal_lm/cpp/chat_model_lm.cpp (hunk 6 above)
Comment (olpipi) [reply text_generation/causal_lm/cpp/chat_model_lm.cpp]: removed

File: text_generation/causal_lm/cpp/CMakeLists.txt (hunk 1 above)
Comment (olpipi) [reply text_generation/causal_lm/cpp/CMakeLists.txt:38]: Can we add id in the next PR?

Comment (ilya-lavrenov) [reply text_generation/causal_lm/cpp/CMakeLists.txt:38]: CI should be done in current PR

File: text_generation/causal_lm/cpp/chat_model_lm.cpp [hunk 9]
```diff
@@ -115,7 +117,10 @@ int main(int argc, char* argv[]) try {
                 }
             }
         }
-        std::cout << "Answer: " << detokenize(detokenizer, answer.tokens) << "\n_______\n";
+
+        auto answer_str = detokenize(detokenizer, answer.tokens);
+        answer_str = answer_str.substr(0, answer_str.find("<eos>"));
```
Comment (ilya-lavrenov) [inline text_generation/causal_lm/cpp/chat_model_lm.cpp:122]: why do we need this hack?
CC @apaniukov

Comment (olpipi) [reply text_generation/causal_lm/cpp/chat_model_lm.cpp:122]: Due to beam search. Some beams may be much shorter than others and they looks like this 
"The answer is 4. 2 + 2 = 4.<eos><eos><eos><eos><eos><eos><eos><eos>..."
I add this just to make output more readable

Comment (apaniukov) [reply text_generation/causal_lm/cpp/chat_model_lm.cpp:122]: No, adding `--skip-special-tokens` to the `convert_tokenizer` command should remove these tokens during detokenization.

Why don't we end the beam generation in the case of the `<eos>` token anyway? It should speed up other beam generation in that case. It also might affect a beam score.

Comment (ilya-lavrenov) [reply text_generation/causal_lm/cpp/chat_model_lm.cpp:122]: > Why don't we end the beam generation in the case of the <eos> token anyway?

It should work this way, I wonder why EOS does not work..

Comment (as-suvorov) [reply text_generation/causal_lm/cpp/chat_model_lm.cpp:122]: It should indeed, I'll check this sample

Comment (as-suvorov) [reply text_generation/causal_lm/cpp/chat_model_lm.cpp:122]: I cannot reproduce it locally. Getting result without eos tokens from detokenizer. I'll sync with @olpipi on that.

File: text_generation/causal_lm/cpp/chat_model_lm.cpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+            if (next_tokens.empty()) {
+                break;
+            }
+            size_t batch_size = next_tokens.size();
+            // Set pointers
+            lm.set_tensor("input_ids", ov::Tensor{ov::element::i64, {batch_size, 1}, next_tokens.data()});
+            lm.set_tensor("beam_idx", ov::Tensor{ov::element::i32, {batch_size}, next_beams.data()});
+            // Set auxiliary inputs
+            ov::Tensor attention_mask = lm.get_tensor("attention_mask");
+            ov::Shape mask_shape{batch_size, attention_mask.get_shape().at(seq_len_dim_idx) + 1};
```
Comment (as-suvorov) [inline text_generation/causal_lm/cpp/chat_model_lm.cpp:103]: It seems mask size should not be increased at the last step. As on the next inference the whole new prompt ids would be passed instead of single token.

**Review by Wovchena — CHANGES_REQUESTED** [review]
